apps/epic-react/src/templates/product-template.tsx

```diff
@@ -20,17 +20,45 @@ import {ProductPageProps} from '@/pages/products/[slug]'
 
 const ProductTemplate: React.FC<ProductPageProps> = ({
   product,
+  products,
   mdx,
   allowPurchase,
   userId,
   couponFromCode,
+  couponIdFromCoupon,
   availableBonuses,
 }) => {
   const router = useRouter()
   const title = product?.title || product?.name || 'Product Title'
 
   const {addPrice} = usePriceCheck()
 
+  const restrictedToProduct = couponFromCode?.restrictedToProductId
+    ? products.find(
+        (product) => product.productId === couponFromCode.restrictedToProductId,
+      )
+    : undefined
+
+  const productMetadata = restrictedToProduct
+    ? {
+        ...restrictedToProduct,
+        id: restrictedToProduct.productId,
+        image: {
+          ...restrictedToProduct.image,
+          width: 132,
+          height: 112,
+        },
+      }
+    : undefined
+
+  const {redeemableCoupon, RedeemDialogForCoupon, validCoupon} = useCoupon(
+    couponFromCode,
+    productMetadata,
+  )
+
+  const couponId =
+    couponIdFromCoupon || (validCoupon ? couponFromCode?.id : undefined)
```

Comment:
should we consider implementing golden ticket provider in this app instead of "whack-a-mole" approach? https://github.com/skillrecordings/products/blob/ddcb80827ce97a2bf2c85325ab13bc949476d462/apps/epic-web/src/hooks/use-golden-ticket.tsx

Reply:
totally agree with you! will create a linear issue for this